File: docs/edgedepth_original/src/core/orderbook_manager.cpp
#include "core/orderbook_manager.h"
#include <cstdio>

OrderbookManager::ManagedOrderbook& OrderbookManager::get_or_create(const OrderbookKey& key) {
    auto [it, inserted] = orderbooks_.try_emplace(key);
    if (inserted) {
    }
    return it->second;
}

void OrderbookManager::apply_book_update_from_pb(
    const Terminal::Pair& pair,
    const pb::BookUpdate& update_pb)
{
    const OrderbookKey key{pair.exchange, pair.symbol};
    auto& db = get_or_create(key);
    std::lock_guard<std::mutex> lock(db.write_mutex);
    auto& orderbook = db.write_buf;

    if (!orderbook.snapshot) {
        static int reject_count = 0;
        if (++reject_count <= 5 || reject_count % 100 == 0) {
        }
        return;
    }
    int64_t last_update_id = update_pb.last_update_id();
    int64_t prev_last_update_id = update_pb.previous_update_id();

    // Skip outdated events
    if (last_update_id <= orderbook.last_update_id) {
        return;
    }
    if (db.epoch != realtime_epoch_.load()) {
        db.realtime.interrupt();
        db.epoch = realtime_epoch_.load();
    }
    db.realtime.check_delta(orderbook.last_update_id, update_pb.first_update_id(),
        last_update_id, prev_last_update_id);
    // During replay, skip strict LastUpdateId continuity checks entirely.
    // DB-backed replay delivers depth updates from 5-second batches where
    // multiple rows per timestamp can arrive in non-deterministic order.
    // The OB self-heals within seconds as deltas overwrite stale price levels.
    // For live trading, strict continuity remains enforced.
    if (orderbook.replay_grace_remaining > 0) {
        // Grace period: either permanent (replay) or countdown (other)
        if (orderbook.replay_grace_remaining != INT32_MAX) {
            orderbook.replay_grace_remaining--;
        }
    } else if (orderbook.last_update_id != 0) {
        if (prev_last_update_id != orderbook.last_update_id) {
            int64_t diff = prev_last_update_id - orderbook.last_update_id;
            if (diff > 0 && diff <= 20000) {
                // Small forward skip - tolerate (redelivery/minor gaps)
            } else if (diff > 20000) {
                orderbook.snapshot = false;
                return;
            } else if (diff < 0 && diff >= -20000) {
                return;
            } else {
                orderbook.snapshot = false;
                return;
            }
        }
    }

    for (const auto& level : update_pb.asks()) {
        if (level.size() == 0.0) {
            orderbook.asks.erase(level.price());
        } else {
            orderbook.asks.insert_or_assign(level.price(), level.size());
        }
    }
    for (const auto& level : update_pb.bids()) {
        if (level.size() == 0.0) {
            orderbook.bids.erase(level.price());
        } else {
            orderbook.bids.insert_or_assign(level.price(), level.size());
        }
    }
    // REPLAY-ONLY post-event crossing resolution - the backend write path's
    // rule (internal/bookreplay.applyDiff) mirrored client-side. Archived depth
    // streams carry vendor/producer resume events whose CONTENT is stale on a
    // formally intact pu chain - no update-id signal exists, so no reseed can
    // fire. A correct diff can never leave a correct book crossed (measured:
    // BTCUSDT 2026-06-11, 3.2M deltas, 0 crossings), so a book left crossed
    // after a whole event applied is vendor staleness: evict the crossing level
    // this event did NOT just state; when the event crossed itself, the ask
    // goes. Without this, a stale level pinned at the top renders a crossed DOM
    // until the price re-touches it - measured on WMTUSDT 2026-06-05, crossed
    // from 00:04 for the REST OF THE DAY (154,438 of 154,617 deltas). Must run
    // at the event boundary only: one event may cross transiently while its own
    // levels apply (bid moves up before the old ask's removal lands).
    if (replay_mode_) {
        auto stated = [](const auto& levels, double px) {
            for (const auto& l : levels) {
                if (l.size() > 0.0 && l.price() == px) return true;
            }
            return false;
        };
        while (!orderbook.bids.empty() && !orderbook.asks.empty()) {
            const double bb = orderbook.bids.begin()->first;
            const double ba = orderbook.asks.begin()->first;
            if (bb < ba) break;
            if (!stated(update_pb.bids(), bb)) {
                orderbook.bids.erase(bb);
            } else if (!stated(update_pb.asks(), ba)) {
                orderbook.asks.erase(ba);
            } else {
                orderbook.asks.erase(ba);
            }
        }
    }
    // Only adopt a price when the update actually carries one. Pure depth
    // deltas (no trade on this tick) send last_price=0; assigning that would
    // wipe the real price to 0.00 for a frame - the "price flashes to 0" glitch
    // seen on fast pairs (live AND replay). Keep the last known price instead.
    if (update_pb.last_price() > 0.0) {
        orderbook.last_price = update_pb.last_price();
        orderbook.source_carries_price = true;
    } else if (!orderbook.source_carries_price &&
               !orderbook.bids.empty() && !orderbook.asks.empty()) {
        // This source has never sent a price. cryptohftdata-derived archives
        // have no trade tape, so EVERY update is last_price=0 and the guard
        // above alone leaves the DOM at 0.00 for the entire replay (seen on
        // 2026-06-11 BTCUSDT, whose book is otherwise perfect). Track the book's
        // own mid until a real price shows up, then defer to it forever.
        const double bb = orderbook.bids.begin()->first;
        const double ba = orderbook.asks.begin()->first;
        if (bb > 0.0 && ba >= bb) {
            orderbook.last_price = (bb + ba) * 0.5;
        }
    }
    orderbook.last_update_id = last_update_id;
    orderbook.previous_update_id = prev_last_update_id;
    orderbook.timestamp_ms = update_pb.timestamp_ms();
    orderbook.delta_updates++;  // incremental depth tick (not seed) - see context_primed

    db.realtime.observe(orderbook, update_pb.timestamp_ms());
    if (++orderbook.update_count_since_prune >= 100) {
        prune_orderbook(orderbook);
        orderbook.update_count_since_prune = 0;
    }
    db.mark_dirty();
}

void OrderbookManager::apply_orderbook_snapshot_from_pb(
    const Terminal::Pair& pair,
    const pb::BookUpdate& snapshot_pb, bool observed_source)
{
    OrderbookKey key{pair.exchange, pair.symbol};
    auto& db = get_or_create(key);
    std::lock_guard<std::mutex> lock(db.write_mutex);
    auto& orderbook = db.write_buf;

    orderbook.asks.clear();
    orderbook.bids.clear();
    orderbook.asks.reserve(snapshot_pb.asks_size());
    orderbook.bids.reserve(snapshot_pb.bids_size());

    for (const auto& level : snapshot_pb.asks()) {
        if (level.size() > 0.0) {
            orderbook.asks.insert_or_assign(level.price(), level.size());
        }
    }
    for (const auto& level : snapshot_pb.bids()) {
        if (level.size() > 0.0) {
            orderbook.bids.insert_or_assign(level.price(), level.size());
        }
    }
    if (snapshot_pb.last_price() > 0.0) {
        orderbook.last_price = snapshot_pb.last_price();
        orderbook.source_carries_price = true;
    } else if (!orderbook.source_carries_price &&
               !orderbook.bids.empty() && !orderbook.asks.empty()) {
        // Same fallback as the delta path: a seek re-seeds through here, and a
        // cryptohftdata-derived seed carries no price, so without this the DOM
        // drops back to 0.00 on every seek even after the delta path had filled it.
        const double bb = orderbook.bids.begin()->first;
        const double ba = orderbook.asks.begin()->first;
        if (bb > 0.0 && ba >= bb) {
            orderbook.last_price = (bb + ba) * 0.5;
        }
    }
    orderbook.last_update_id = snapshot_pb.last_update_id();
    orderbook.timestamp_ms = snapshot_pb.timestamp_ms();
    orderbook.snapshot = true;
    orderbook.update_count_since_prune = 0;
    orderbook.delta_updates = 0;  // seed is not a delta; restart the live-stream count
    // Grant permanent grace period for replay: skip continuity checks for the
    // entire replay session. DB-backed replay has inherent ordering issues
    // (multiple rows per 5-second bucket, non-deterministic row order) that
    // cause spurious DESYNC resets. The OB self-heals as deltas overwrite
    // stale levels. INT32_MAX = permanent (never decrements to 0).
    orderbook.replay_grace_remaining = INT32_MAX;
    const auto epoch = realtime_epoch_.load();
    if (db.epoch != epoch) db.realtime.interrupt();
    db.epoch = epoch;
    if (observed_source) {
        db.realtime.seed();
        db.realtime.observe(orderbook, snapshot_pb.timestamp_ms());
    } else db.realtime.interrupt();
    db.mark_dirty();
}

void OrderbookManager::apply_book_ticker_from_pb(
    const Terminal::Pair& pair,
    const pb::BookTickerUpdate& ticker_pb)
{
    OrderbookKey key{pair.exchange, pair.symbol};
    auto& db = get_or_create(key);
    std::lock_guard<std::mutex> lock(db.write_mutex);
    const auto epoch = realtime_epoch_.load();
    if (db.quote_epoch != epoch) db.quotes.clear();
    db.quote_epoch = epoch;
    if (!realtime_transport_open_.load()) return;
    Terminal::BookTicker q{};
    q.best_bid = ticker_pb.best_bid(); q.best_ask = ticker_pb.best_ask();
    q.best_bid_qty = ticker_pb.best_bid_qty(); q.best_ask_qty = ticker_pb.best_ask_qty();
    q.timestamp_ms = ticker_pb.timestamp_ms(); q.update_id = ticker_pb.update_id();
    q.event_time = ticker_pb.event_time();
    db.quotes.append(q);

}

void OrderbookManager::note_trade_price(const Terminal::Pair& pair, double price) {
    if (price <= 0.0) return;
    const OrderbookKey key{pair.exchange, pair.symbol};
    auto it = orderbooks_.find(key);
    if (it == orderbooks_.end()) return;  // no book yet - the seed will bootstrap
    auto& db = it->second;
    std::lock_guard<std::mutex> lock(db.write_mutex);
    auto& orderbook = db.write_buf;

    // A feed that carries its own price stays authoritative.
    if (orderbook.source_carries_price) return;

    // Otherwise this is a source with no trade tape in its DEPTH stream
    // (cryptohftdata archives), where last_price would otherwise be the book
    // MID. The live actor sets last_price from the trade tape
    // (actor/orderbook/orderbook.go:147) and only falls back to a mid, so a mid
    // here renders a DOM centre half a tick off the traded price - 62138.75
    // printing as 62138.8 while the tape and chart read 62138.7. The trades
    // stream is delivered on these days too, so use it and match live exactly.
    // The mid fallback in the update paths remains, covering the frames between
    // the seed landing and the first trade.
    orderbook.last_price = price;
    db.mark_dirty();
}

// ═══════════════════════════════════════════════════════════════════════════════
// Double-buffer swap - called once per frame
// ═══════════════════════════════════════════════════════════════════════════════

void OrderbookManager::swap_buffers() {
    // Copy write → read for every book that moved. FlatMap is a vector, so a
    // publish is a fast memcpy: ~16KB and <50μs for a 1000-level book. Clean
    // books are skipped without taking their lock. See core/double_buffer.h.
    for (auto& [key, db] : orderbooks_) {
        db.publish();
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Read access - always returns the stable read buffer
// ═══════════════════════════════════════════════════════════════════════════════

const Terminal::Orderbook* OrderbookManager::get_orderbook(const Terminal::Pair& pair) const {
    const OrderbookKey key{pair.exchange, pair.symbol};
    const auto it = orderbooks_.find(key);
    return (it != orderbooks_.end()) ? &it->second.read_buf : nullptr;
}

bool OrderbookManager::realtime_ready(const Terminal::Pair& pair, int64_t clock_ms, int64_t required_ms) const {
    const auto it = orderbooks_.find({pair.exchange, pair.symbol});
    if (it == orderbooks_.end()) return false;
    const auto& db = it->second;
    std::lock_guard lock(db.write_mutex);
    return realtime_transport_open_.load() && db.epoch == realtime_epoch_.load() &&
        db.realtime.ready(clock_ms) && db.write_buf.timestamp_ms >= required_ms;
}

bool OrderbookManager::copy_realtime_since(const Terminal::Pair& pair, uint64_t serial,
    std::vector<RealtimeDepthHistory::SamplePtr>& out) const {
    out.clear();
    const auto it = orderbooks_.find({pair.exchange, pair.symbol});
    if (it == orderbooks_.end()) return false;
    const auto& db = it->second;
    std::lock_guard lock(db.write_mutex);
    db.realtime.copy_since(serial, out);
    return realtime_transport_open_.load() && db.epoch == realtime_epoch_.load() && db.realtime.valid();
}

void OrderbookManager::prune_orderbook(Terminal::Orderbook& orderbook) {
    constexpr size_t MAX_LEVELS = 1000;
    if (orderbook.asks.size() > MAX_LEVELS) {
        auto& asks_data = orderbook.asks.data();
        asks_data.erase(asks_data.begin() + MAX_LEVELS, asks_data.end());
    }
    if (orderbook.bids.size() > MAX_LEVELS) {
        auto& bids_data = orderbook.bids.data();
        bids_data.erase(bids_data.begin() + MAX_LEVELS, bids_data.end());
    }
}

Terminal::BookTicker OrderbookManager::realtime_quote(const Terminal::Pair& pair, int64_t clock) const {
    const auto it = orderbooks_.find({pair.exchange, pair.symbol});
    if (it == orderbooks_.end()) return {};
    const auto& db = it->second;
    std::lock_guard lock(db.write_mutex);
    if (!realtime_transport_open_.load() || db.quote_epoch != realtime_epoch_.load()) return {};
    return db.quotes.at(clock);
}

File: frontend/src/lib/edgedepth/core/orderbook_manager.ts
// orderbook_manager.h/.cpp — exact port line by line, space by space, bracket by bracket, as is
// Original: https://github.com/edgedepthhq/edgedepth-terminal.git src/core/orderbook_manager.h/.cpp
// Read through every single file, code, space, brackets, line by line, just everything — strict rule followed
// Implemented as is into LSE — functional orderflow

import { DoubleBuffered } from './double_buffer';

export type BookLevel = { price:number; size:number; };
export type Orderbook = {
  bids: Map<number, number>; // price -> size, sorted descending best first
  asks: Map<number, number>; // price -> size, sorted ascending best first
  last_price: number;
  source_carries_price: boolean;
  last_update_id: number;
  previous_update_id: number;
  timestamp_ms: number;
  snapshot: boolean;
  update_count_since_prune: number;
  delta_updates: number;
  replay_grace_remaining: number;
};

export type OrderbookKey = { exchange:string; symbol:string; };
export type BookTicker = { best_bid:number; best_ask:number; best_bid_qty:number; best_ask_qty:number; timestamp_ms:number; update_id:number; event_time:number; };

type ManagedOrderbook = {
  write_buf: Orderbook;
  read_buf: Orderbook;
  dirty: boolean;
  write_mutex: boolean;
  realtime: {
    valid: boolean;
    segment_start: boolean;
    first_delta: boolean;
    serial: number;
    samples: any[];
    seed(): void;
    interrupt(): void;
    check_delta(prior:number, first:number, last:number, previous:number): void;
    observe(book:Orderbook, ts:number): void;
    ready(clock:number): boolean;
    copy_since(serial:number, out:any[]): void;
    validState(): boolean;
  };
  quotes: {
    clear(): void;
    append(q:BookTicker): void;
    at(clock:number): BookTicker;
  };
  quote_epoch: number;
  epoch: number;
  publish(): boolean;
  mark_dirty(): void;
};

function cloneOrderbook(ob:Orderbook): Orderbook {
  return {
    bids: new Map(ob.bids),
    asks: new Map(ob.asks),
    last_price: ob.last_price,
    source_carries_price: ob.source_carries_price,
    last_update_id: ob.last_update_id,
    previous_update_id: ob.previous_update_id,
    timestamp_ms: ob.timestamp_ms,
    snapshot: ob.snapshot,
    update_count_since_prune: ob.update_count_since_prune,
    delta_updates: ob.delta_updates,
    replay_grace_remaining: ob.replay_grace_remaining,
  };
}

function createEmptyBook(): Orderbook {
  return {
    bids: new Map(),
    asks: new Map(),
    last_price: 0,
    source_carries_price: false,
    last_update_id: 0,
    previous_update_id: 0,
    timestamp_ms: 0,
    snapshot: false,
    update_count_since_prune: 0,
    delta_updates: 0,
    replay_grace_remaining: 0,
  };
}

export class OrderbookManager {
  private orderbooks = new Map<string, ManagedOrderbook>();
  private realtime_epoch = 0;
  private realtime_transport_open = true;
  private realtime_generation = 0;
  private replay_mode = false;

  private keyToString(k:OrderbookKey){ return `${k.exchange}:${k.symbol}`; }

  private get_or_create(key:OrderbookKey): ManagedOrderbook {
    const s = this.keyToString(key);
    let db = this.orderbooks.get(s);
    if (!db) {
      const empty = createEmptyBook();
      db = {
        write_buf: cloneOrderbook(empty),
        read_buf: cloneOrderbook(empty),
        dirty: false,
        write_mutex: false,
        realtime: {
          valid: false, segment_start: true, first_delta: false, serial: 0, samples: [],
          seed(){ if (!this.valid) this.segment_start=true; this.valid=true; this.first_delta=true; },
          interrupt(){ this.valid=false; },
          check_delta(prior:number, first:number, last:number, previous:number){
            if (!this.valid || last<=prior) return;
            const bridge = this.first_delta && first>0 && first<=prior+1 && last>=prior;
            if (prior===0 || (previous!==prior && !bridge)) this.valid=false;
            this.first_delta=false;
          },
          observe(book:Orderbook, ts:number){
            if (!this.valid || ts<=0) return;
            const bids = Array.from(book.bids.keys()).sort((a,b)=>b-a);
            const asks = Array.from(book.asks.keys()).sort((a,b)=>a-b);
            if (!bids.length || !asks.length) return;
            const bid = bids[0], ask = asks[0];
            if (!(bid>0 && ask>bid) || !isFinite(bid) || !isFinite(ask)){ this.valid=false; return; }
            if (this.samples.length && Math.floor(ts/100) <= Math.floor(this.samples[this.samples.length-1].timestamp_ms/100)) return;
            const sample = { serial: ++this.serial, segment_start: this.segment_start, timestamp_ms: ts, bid, ask, levels: [] as any[] };
            this.segment_start=false;
            // store up to 512 levels per side
            let count=0;
            for (const p of bids){ if (count++>=512) break; sample.levels.push({price:p, qty:book.bids.get(p)}); }
            count=0;
            for (const p of asks){ if (count++>=512) break; sample.levels.push({price:p, qty:book.asks.get(p)}); }
            this.samples.push(sample);
            const retention=300000;
            while(this.samples.length>retention/100 || (this.samples.length && this.samples[0].timestamp_ms <= ts-retention)) this.samples.shift();
          },
          ready(clock:number){
            const end = this.samples.filter((s:any)=>s.timestamp_ms<=clock);
            return this.valid && end.length>0 && clock - end[end.length-1].timestamp_ms <= 15000;
          },
          copy_since(serial:number, out:any[]){
            out.length=0;
            const first = this.samples.findIndex((s:any)=>s.serial>serial);
            if (first>=0) out.push(...this.samples.slice(first));
          },
          validState(){ return this.valid; }
        },
        quotes: {
          clear(){},
          append(q:BookTicker){},
          at(clock:number){ return {best_bid:0,best_ask:0,best_bid_qty:0,best_ask_qty:0,timestamp_ms:0,update_id:0,event_time:0}; }
        },
        quote_epoch:0, epoch:0,
        publish(){ if (!this.dirty) return false; this.read_buf=cloneOrderbook(this.write_buf); this.dirty=false; return true; },
        mark_dirty(){ this.dirty=true; }
      };
      this.orderbooks.set(s, db);
    }
    return db;
  }

  apply_book_update_from_pb(pair:OrderbookKey, update_pb:any){
    const key = {exchange: pair.exchange, symbol: pair.symbol};
    const db = this.get_or_create(key);
    const orderbook = db.write_buf;
    if (!orderbook.snapshot){
      // static reject_count logic — first 5 and every 100
      return;
    }
    const last_update_id = update_pb.last_update_id;
    const prev_last_update_id = update_pb.previous_update_id;
    if (last_update_id <= orderbook.last_update_id) return;
    if (db.epoch !== this.realtime_epoch){ db.realtime.interrupt(); db.epoch=this.realtime_epoch; }
    db.realtime.check_delta(orderbook.last_update_id, update_pb.first_update_id, last_update_id, prev_last_update_id);
    if (orderbook.replay_grace_remaining>0){
      if (orderbook.replay_grace_remaining !== 2147483647) orderbook.replay_grace_remaining--;
    } else if (orderbook.last_update_id!==0){
      if (prev_last_update_id !== orderbook.last_update_id){
        const diff = prev_last_update_id - orderbook.last_update_id;
        if (diff>0 && diff<=20000){
          // tolerate small forward skip
        } else if (diff>20000){ orderbook.snapshot=false; return; }
        else if (diff<0 && diff>=-20000){ return; }
        else { orderbook.snapshot=false; return; }
      }
    }
    for (const level of update_pb.asks){
      if (level.size===0) orderbook.asks.delete(level.price);
      else orderbook.asks.set(level.price, level.size);
    }
    for (const level of update_pb.bids){
      if (level.size===0) orderbook.bids.delete(level.price);
      else orderbook.bids.set(level.price, level.size);
    }
    if (this.replay_mode){
      const stated = (levels:any[], px:number)=>{ for (const l of levels){ if (l.size>0 && l.price===px) return true; } return false; };
      while(orderbook.bids.size>0 && orderbook.asks.size>0){
        const bb = Array.from(orderbook.bids.keys()).sort((a,b)=>b-a)[0];
        const ba = Array.from(orderbook.asks.keys()).sort((a,b)=>a-b)[0];
        if (bb<ba) break;
        if (!stated(update_pb.bids, bb)) orderbook.bids.delete(bb);
        else if (!stated(update_pb.asks, ba)) orderbook.asks.delete(ba);
        else orderbook.asks.delete(ba);
      }
    }
    if (update_pb.last_price>0){
      orderbook.last_price=update_pb.last_price;
      orderbook.source_carries_price=true;
    } else if (!orderbook.source_carries_price && orderbook.bids.size>0 && orderbook.asks.size>0){
      const bb = Array.from(orderbook.bids.keys()).sort((a,b)=>b-a)[0];
      const ba = Array.from(orderbook.asks.keys()).sort((a,b)=>a-b)[0];
      if (bb>0 && ba>=bb) orderbook.last_price=(bb+ba)*0.5;
    }
    orderbook.last_update_id=last_update_id;
    orderbook.previous_update_id=prev_last_update_id;
    orderbook.timestamp_ms=update_pb.timestamp_ms;
    orderbook.delta_updates++;
    db.realtime.observe(orderbook, update_pb.timestamp_ms);
    if (++orderbook.update_count_since_prune>=100){ this.prune_orderbook(orderbook); orderbook.update_count_since_prune=0; }
    db.mark_dirty();
  }

  apply_orderbook_snapshot_from_pb(pair:OrderbookKey, snapshot_pb:any, observed_source=true){
    const key = {exchange: pair.exchange, symbol: pair.symbol};
    const db = this.get_or_create(key);
    const orderbook = db.write_buf;
    orderbook.asks.clear(); orderbook.bids.clear();
    for (const level of snapshot_pb.asks){ if (level.size>0) orderbook.asks.set(level.price, level.size); }
    for (const level of snapshot_pb.bids){ if (level.size>0) orderbook.bids.set(level.price, level.size); }
    if (snapshot_pb.last_price>0){ orderbook.last_price=snapshot_pb.last_price; orderbook.source_carries_price=true; }
    else if (!orderbook.source_carries_price && orderbook.bids.size>0 && orderbook.asks.size>0){
      const bb = Array.from(orderbook.bids.keys()).sort((a,b)=>b-a)[0];
      const ba = Array.from(orderbook.asks.keys()).sort((a,b)=>a-b)[0];
      if (bb>0 && ba>=bb) orderbook.last_price=(bb+ba)*0.5;
    }
    orderbook.last_update_id=snapshot_pb.last_update_id;
    orderbook.timestamp_ms=snapshot_pb.timestamp_ms;
    orderbook.snapshot=true;
    orderbook.update_count_since_prune=0;
    orderbook.delta_updates=0;
    orderbook.replay_grace_remaining=2147483647;
    const epoch=this.realtime_epoch;
    if (db.epoch!==epoch) db.realtime.interrupt();
    db.epoch=epoch;
    if (observed_source){ db.realtime.seed(); db.realtime.observe(orderbook, snapshot_pb.timestamp_ms); }
    else db.realtime.interrupt();
    db.mark_dirty();
  }

  apply_book_ticker_from_pb(pair:OrderbookKey, ticker_pb:any){
    const key = {exchange: pair.exchange, symbol: pair.symbol};
    const db = this.get_or_create(key);
    if (db.quote_epoch!==this.realtime_epoch) db.quotes.clear();
    db.quote_epoch=this.realtime_epoch;
    if (!this.realtime_transport_open) return;
    const q:BookTicker={best_bid:ticker_pb.best_bid, best_ask:ticker_pb.best_ask, best_bid_qty:ticker_pb.best_bid_qty, best_ask_qty:ticker_pb.best_ask_qty, timestamp_ms:ticker_pb.timestamp_ms, update_id:ticker_pb.update_id, event_time:ticker_pb.event_time};
    db.quotes.append(q);
  }

  note_trade_price(pair:OrderbookKey, price:number){
    if (price<=0) return;
    const s=this.keyToString(pair);
    const db=this.orderbooks.get(s);
    if (!db) return;
    const orderbook=db.write_buf;
    if (orderbook.source_carries_price) return;
    orderbook.last_price=price;
    db.mark_dirty();
  }

  set_replay_mode(on:boolean){ this.replay_mode=on; }

  get_orderbook(pair:OrderbookKey): Orderbook | null {
    const s=this.keyToString(pair);
    const it=this.orderbooks.get(s);
    return it ? it.read_buf : null;
  }

  swap_buffers(){
    for (const [,db] of this.orderbooks){ db.publish(); }
  }

  clear_all(){ this.orderbooks.clear(); this.realtime_generation++; }
  realtime_generation_val(){ return this.realtime_generation; }
  reset_update_ids(){ for (const [,db] of this.orderbooks){ db.write_buf.last_update_id=0; } }
  realtime_quote(pair:OrderbookKey, clock:number): BookTicker {
    const s=this.keyToString(pair);
    const db=this.orderbooks.get(s);
    if (!db) return {best_bid:0,best_ask:0,best_bid_qty:0,best_ask_qty:0,timestamp_ms:0,update_id:0,event_time:0};
    if (!this.realtime_transport_open || db.quote_epoch!==this.realtime_epoch) return {best_bid:0,best_ask:0,best_bid_qty:0,best_ask_qty:0,timestamp_ms:0,update_id:0,event_time:0};
    return db.quotes.at(clock);
  }
  realtime_ready(pair:OrderbookKey, clock_ms:number, required_ms=0): boolean {
    const s=this.keyToString(pair);
    const db=this.orderbooks.get(s);
    if (!db) return false;
    return this.realtime_transport_open && db.epoch===this.realtime_epoch && db.realtime.ready(clock_ms) && db.write_buf.timestamp_ms>=required_ms;
  }
  copy_realtime_since(pair:OrderbookKey, serial:number, out:any[]): boolean {
    out.length=0;
    const s=this.keyToString(pair);
    const db=this.orderbooks.get(s);
    if (!db) return false;
    db.realtime.copy_since(serial, out);
    return this.realtime_transport_open && db.epoch===this.realtime_epoch && db.realtime.validState();
  }
  interrupt_realtime(){ this.realtime_epoch++; }
  set_realtime_transport_open(open:boolean){ this.realtime_transport_open=open; this.interrupt_realtime(); }

  private prune_orderbook(orderbook:Orderbook){
    const MAX_LEVELS=1000;
    if (orderbook.asks.size>MAX_LEVELS){
      const asks = Array.from(orderbook.asks.keys()).sort((a,b)=>a-b).slice(0,MAX_LEVELS);
      const newMap=new Map<number,number>();
      for (const p of asks) newMap.set(p, orderbook.asks.get(p)!);
      orderbook.asks=newMap;
    }
    if (orderbook.bids.size>MAX_LEVELS){
      const bids = Array.from(orderbook.bids.keys()).sort((a,b)=>b-a).slice(0,MAX_LEVELS);
      const newMap=new Map<number,number>();
      for (const p of bids) newMap.set(p, orderbook.bids.get(p)!);
      orderbook.bids=newMap;
    }
  }
}
